perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/VosManagerBlImpl.java

```diff
@@ -902,14 +929,65 @@ public boolean usesEmbeddedGroupRegistrations(PerunSession sess, Vo vo) {
 
 	@Override
 	public void addMemberVo(PerunSession sess, Vo vo, Vo memberVo) throws RelationExistsException {
-		// todo - add necessary logic for adding new member vo
 		vosManagerImpl.addMemberVo(sess, vo, memberVo);
+
+		List<Member> memberVoMembers = perunBl.getMembersManagerBl().getMembers(sess, memberVo);
+		AttributeDefinition memberExpirationAttrDef;
+
+		try {
+			memberExpirationAttrDef = perunBl.getAttributesManagerBl().getAttributeDefinition(sess, AttributesManager.NS_MEMBER_ATTR_DEF + ":membershipExpiration");
+		} catch (AttributeNotExistsException e) {
+			throw new InternalErrorException(e);
+		}
+
+		for (Member member : memberVoMembers) {
+			try {
+
+				try {
+					// if user is member in both vos, update memberOrganizations list attribute, validate and reset expiration
+					Member existingMember = perunBl.getMembersManagerBl().getMemberByUserId(sess, vo, member.getUserId());
+					perunBl.getAttributesManagerBl().removeAttribute(sess, existingMember, memberExpirationAttrDef);
+					if (existingMember.getStatus().equals(Status.INVALID) || existingMember.getStatus().equals(Status.DISABLED)) {
```

Comment:
We should probably validate existing members if they are expired as well, shouldn't we?

Reply:
I'm not sure, expired members should already have validated attributes, so we'd set the status from EXPIRED to VALID and run the attributes validation check again, but wouldn't the status get changed to VALID after adding the new member vo to memberOrganizations anyway?